File: logica/lexer.py
"""
Logica Lexer — Breaking Source Into Tokens
==========================================

Twenty-five keywords. Three values. One language.
"""

from enum import Enum, auto
from dataclasses import dataclass
from typing import Optional
from .errors import LexError


class TokenType(Enum):
    # Literals
    INTEGER = auto()
    FLOAT = auto()
    STRING = auto()
    BOOLEAN = auto()

    # Identifiers
    IDENTIFIER = auto()

    # Keywords — Identity
    SPEAKER = auto()       # speaker
    AS = auto()            # as

    # Keywords — Variables
    LET = auto()           # let
    READ = auto()          # read

    # Keywords — Expressions
    SPEAK = auto()         # speak
    WHEN = auto()           # when
    OTHERWISE = auto()     # otherwise
    BROKEN = auto()        # broken

    # Keywords — Functions
    FN = auto()            # fn
    RETURN = auto()        # return

    # Keywords — Loops
    WHILE = auto()         # while
    MAX = auto()           # max

    # Keywords — Communication
    REQUEST = auto()       # request
    RESPOND = auto()       # respond
    ACCEPT = auto()        # accept
    REFUSE = auto()        # refuse

    # Keywords — Inspection
    INSPECT = auto()       # inspect
    HISTORY = auto()       # history
    LEDGER = auto()        # ledger
    VERIFY = auto()        # verify

    # Keywords — Worlds
    WORLD = auto()         # world
    SEAL = auto()          # seal

    # Keywords — Logic
    AND = auto()           # and
    OR = auto()            # or
    NOT = auto()           # not

    # Keywords — Values
    ACTIVE = auto()        # active
    INACTIVE = auto()      # inactive
    TRUE = auto()          # true
    FALSE = auto()         # false
    NONE = auto()          # none

    # Keywords — Control
    IF = auto()            # if
    ELIF = auto()          # elif
    ELSE = auto()          # else
    PASS = auto()          # pass
    FAIL = auto()          # fail

    # Operators
    PLUS = auto()          # +
    MINUS = auto()         # -
    STAR = auto()          # *
    SLASH = auto()         # /
    PERCENT = auto()       # %
    ASSIGN = auto()        # =
    EQ = auto()            # ==
    NEQ = auto()           # !=
    LT = auto()            # <
    GT = auto()            # >
    LTE = auto()           # <=
    GTE = auto()           # >=
    DOT = auto()           # .
    COMMA = auto()         # ,
    COLON = auto()         # :
    ARROW = auto()         # ->

    # Delimiters
    LBRACE = auto()        # {
    RBRACE = auto()        # }
    LPAREN = auto()        # (
    RPAREN = auto()        # )
    LBRACKET = auto()      # [
    RBRACKET = auto()      # ]

    # Special
    NEWLINE = auto()
    EOF = auto()
    COMMENT = auto()


# Keyword map
KEYWORDS = {
    "speaker": TokenType.SPEAKER,
    "as": TokenType.AS,
    "let": TokenType.LET,
    "read": TokenType.READ,
    "speak": TokenType.SPEAK,
    "when": TokenType.WHEN,
    "otherwise": TokenType.OTHERWISE,
    "broken": TokenType.BROKEN,
    "fn": TokenType.FN,
    "return": TokenType.RETURN,
    "while": TokenType.WHILE,
    "max": TokenType.MAX,
    "request": TokenType.REQUEST,
    "respond": TokenType.RESPOND,
    "accept": TokenType.ACCEPT,
    "refuse": TokenType.REFUSE,
    "inspect": TokenType.INSPECT,
    "history": TokenType.HISTORY,
    "ledger": TokenType.LEDGER,
    "verify": TokenType.VERIFY,
    "world": TokenType.WORLD,
    "seal": TokenType.SEAL,
    "and": TokenType.AND,
    "or": TokenType.OR,
    "not": TokenType.NOT,
    "active": TokenType.ACTIVE,
    "inactive": TokenType.INACTIVE,
    "true": TokenType.TRUE,
    "false": TokenType.FALSE,
    "none": TokenType.NONE,
    "if": TokenType.IF,
    "elif": TokenType.ELIF,
    "else": TokenType.ELSE,
    "pass": TokenType.PASS,
    "fail": TokenType.FAIL,
}


@dataclass
class Token:
    type: TokenType
    value: str
    line: int
    col: int

    def __repr__(self):
        return f"Token({self.type.name}, {self.value!r}, L{self.line}:{self.col})"


class Lexer:
    """
    Breaks Logica source code into tokens.

    Supports:
    - # line comments
    - String literals with " or '
    - Integer and float literals
    - All operators and delimiters
    - 25 keywords
    """

    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.col = 1
        self.tokens: list[Token] = []

    def tokenize(self) -> list[Token]:
        """Tokenize the entire source. Returns list of tokens."""
        while self.pos < len(self.source):
            self._skip_whitespace()
            if self.pos >= len(self.source):
                break

            ch = self.source[self.pos]

            # Newlines
            if ch == '\n':
                self._add(TokenType.NEWLINE, '\n')
                self._advance()
                continue

            # Comments
            if ch == '#':
                self._read_comment()
                continue

            # Strings
            if ch in ('"', "'"):
                self._read_string(ch)
                continue

            # Numbers
            if ch.isdigit():
                self._read_number()
                continue

            # Identifiers and keywords
            if ch.isalpha() or ch == '_':
                self._read_identifier()
                continue

            # Two-character operators
            if self.pos + 1 < len(self.source):
                two = self.source[self.pos:self.pos + 2]
                if two == '==':
                    self._add(TokenType.EQ, '==')
                    self._advance(2)
                    continue
                if two == '!=':
                    self._add(TokenType.NEQ, '!=')
                    self._advance(2)
                    continue
                if two == '<=':
                    self._add(TokenType.LTE, '<=')
                    self._advance(2)
                    continue
                if two == '>=':
                    self._add(TokenType.GTE, '>=')
                    self._advance(2)
                    continue
                if two == '->':
                    self._add(TokenType.ARROW, '->')
                    self._advance(2)
                    continue

            # Single-character operators and delimiters
            single = {
                '+': TokenType.PLUS,
                '-': TokenType.MINUS,
                '*': TokenType.STAR,
                '/': TokenType.SLASH,
                '%': TokenType.PERCENT,
                '=': TokenType.ASSIGN,
                '<': TokenType.LT,
                '>': TokenType.GT,
                '.': TokenType.DOT,
                ',': TokenType.COMMA,
                ':': TokenType.COLON,
                '{': TokenType.LBRACE,
                '}': TokenType.RBRACE,
                '(': TokenType.LPAREN,
                ')': TokenType.RPAREN,
                '[': TokenType.LBRACKET,
                ']': TokenType.RBRACKET,
            }

            if ch in single:
                self._add(single[ch], ch)
                self._advance()
                continue

            raise LexError(f"unexpected character: {ch!r}", self.line, self.col)

        self._add(TokenType.EOF, '')
        return self.tokens

    def _skip_whitespace(self):
        """Skip spaces and tabs (not newlines)."""
        while self.pos < len(self.source) and self.source[self.pos] in (' ', '\t', '\r'):
            self._advance()

    def _advance(self, n=1):
        """Advance position by n characters."""
        for _ in range(n):
            if self.pos < len(self.source):
                if self.source[self.pos] == '\n':
                    self.line += 1
                    self.col = 1
                else:
                    self.col += 1
                self.pos += 1

    def _add(self, token_type: TokenType, value: str):
        """Add a token to the list."""
        self.tokens.append(Token(token_type, value, self.line, self.col))

    def _read_comment(self):
        """Read a # comment to end of line."""
        start_col = self.col
        comment = ''
        self._advance()  # skip #
        while self.pos < len(self.source) and self.source[self.pos] != '\n':
            comment += self.source[self.pos]
            self._advance()
        # Comments are skipped, not added to tokens

    def _read_string(self, quote: str):
        """Read a string literal."""
        start_line = self.line
        start_col = self.col
        self._advance()  # skip opening quote
        value = ''
        while self.pos < len(self.source):
            ch = self.source[self.pos]
            if ch == '\\' and self.pos + 1 < len(self.source):
                next_ch = self.source[self.pos + 1]
                escapes = {'n': '\n', 't': '\t', '\\': '\\', quote: quote}
                if next_ch in escapes:
                    value += escapes[next_ch]
                    self._advance(2)
                    continue
            if ch == quote:
                self._advance()  # skip closing quote
                self.tokens.append(Token(TokenType.STRING, value, start_line, start_col))
                return
            if ch == '\n':
                raise LexError("unterminated string", start_line, start_col)
            value += ch
            self._advance()
        raise LexError("unterminated string", start_line, start_col)

    def _read_number(self):
        """Read an integer or float literal."""
        start_col = self.col
        value = ''
        is_float = False
        while self.pos < len(self.source) and (self.source[self.pos].isdigit() or self.source[self.pos] == '.'):
            if self.source[self.pos] == '.':
                # Check if this is a decimal point or member access
                if is_float:
                    break  # second dot — stop
                if self.pos + 1 < len(self.source) and self.source[self.pos + 1].isdigit():
                    is_float = True
                else:
                    break  # dot followed by non-digit — member access
            value += self.source[self.pos]
            self._advance()
        token_type = TokenType.FLOAT if is_float else TokenType.INTEGER
        self.tokens.append(Token(token_type, value, self.line, start_col))

    def _read_identifier(self):
        """Read an identifier or keyword."""
        start_col = self.col
        value = ''
        while self.pos < len(self.source) and (self.source[self.pos].isalnum() or self.source[self.pos] == '_'):
            value += self.source[self.pos]
            self._advance()

        # Check for keywords
        token_type = KEYWORDS.get(value, TokenType.IDENTIFIER)
        self.tokens.append(Token(token_type, value, self.line, start_col))

File: HumanLogicaIDE/Sources/HumanLogicaCore/Lexer.swift
// Lexer.swift — Breaking Source Into Tokens
// Supports: # line comments, string literals, integers, floats,
// all operators and delimiters, 25 keywords.

import Foundation

public class Lexer {
    private let source: [Character]
    private var pos: Int = 0
    private var line: Int = 1
    private var col: Int = 1
    public private(set) var tokens: [Token] = []

    public init(source: String) {
        self.source = Array(source)
    }

    /// Tokenize the entire source. Returns list of tokens.
    public func tokenize() throws -> [Token] {
        while pos < source.count {
            skipWhitespace()
            guard pos < source.count else { break }

            let ch = source[pos]

            // Newlines
            if ch == "\n" {
                addToken(.newline, "\n")
                advance()
                continue
            }

            // Comments
            if ch == "#" {
                readComment()
                continue
            }

            // Strings
            if ch == "\"" || ch == "'" {
                try readString(quote: ch)
                continue
            }

            // Numbers
            if ch.isNumber {
                readNumber()
                continue
            }

            // Identifiers and keywords
            if ch.isLetter || ch == "_" {
                readIdentifier()
                continue
            }

            // Two-character operators
            if pos + 1 < source.count {
                let two = String(source[pos...pos+1])
                switch two {
                case "==":
                    addToken(.eq, "=="); advance(2); continue
                case "!=":
                    addToken(.neq, "!="); advance(2); continue
                case "<=":
                    addToken(.lte, "<="); advance(2); continue
                case ">=":
                    addToken(.gte, ">="); advance(2); continue
                case "->":
                    addToken(.arrow, "->"); advance(2); continue
                default:
                    break
                }
            }

            // Single-character operators and delimiters
            let singleMap: [Character: TokenType] = [
                "+": .plus, "-": .minus, "*": .star, "/": .slash,
                "%": .percent, "=": .assign, "<": .lt, ">": .gt,
                ".": .dot, ",": .comma, ":": .colon,
                "{": .lbrace, "}": .rbrace, "(": .lparen, ")": .rparen,
                "[": .lbracket, "]": .rbracket,
            ]

            if let tokenType = singleMap[ch] {
                addToken(tokenType, String(ch))
                advance()
                continue
            }

            throw LogicaError.lexError(message: "unexpected character: '\(ch)'", line: line, col: col)
        }

        addToken(.eof, "")
        return tokens
    }

    // MARK: - Private Helpers

    private func skipWhitespace() {
        while pos < source.count && (source[pos] == " " || source[pos] == "\t" || source[pos] == "\r") {
            advance()
        }
    }

    private func advance(_ n: Int = 1) {
        for _ in 0..<n {
            guard pos < source.count else { return }
            if source[pos] == "\n" {
                line += 1
                col = 1
            } else {
                col += 1
            }
            pos += 1
        }
    }

    private func addToken(_ type: TokenType, _ value: String) {
        tokens.append(Token(type: type, value: value, line: line, col: col))
    }

    private func readComment() {
        advance() // skip #
        while pos < source.count && source[pos] != "\n" {
            advance()
        }
    }

    private func readString(quote: Character) throws {
        let startLine = line
        let startCol = col
        advance() // skip opening quote
        var value = ""

        while pos < source.count {
            let ch = source[pos]
            if ch == "\\" && pos + 1 < source.count {
                let next = source[pos + 1]
                let escapes: [Character: Character] = ["n": "\n", "t": "\t", "\\": "\\", quote: quote]
                if let escaped = escapes[next] {
                    value.append(escaped)
                    advance(2)
                    continue
                }
            }
            if ch == quote {
                advance() // skip closing quote
                tokens.append(Token(type: .string, value: value, line: startLine, col: startCol))
                return
            }
            if ch == "\n" {
                throw LogicaError.lexError(message: "unterminated string", line: startLine, col: startCol)
            }
            value.append(ch)
            advance()
        }
        throw LogicaError.lexError(message: "unterminated string", line: startLine, col: startCol)
    }

    private func readNumber() {
        let startCol = col
        var value = ""
        var isFloat = false

        while pos < source.count && (source[pos].isNumber || source[pos] == ".") {
            if source[pos] == "." {
                if isFloat { break }
                if pos + 1 < source.count && source[pos + 1].isNumber {
                    isFloat = true
                } else {
                    break
                }
            }
            value.append(source[pos])
            advance()
        }

        tokens.append(Token(type: isFloat ? .float : .integer, value: value, line: line, col: startCol))
    }

    private func readIdentifier() {
        let startCol = col
        var value = ""

        while pos < source.count && (source[pos].isLetter || source[pos].isNumber || source[pos] == "_") {
            value.append(source[pos])
            advance()
        }

        let tokenType = logicaKeywords[value] ?? .identifier
        tokens.append(Token(type: tokenType, value: value, line: line, col: startCol))
    }
}
